pkg/controllers/dnspolicy/dnspolicy_healthchecks.go

```diff
@@ -2,203 +2,164 @@ package dnspolicy
+	log := crlog.FromContext(ctx)
+	var healthChecks []*v1alpha1.DNSHealthCheckProbe
+	if dnsPolicy.Spec.HealthCheck == nil {
+		log.V(3).Info("DNS Policy has no defined health check")
+		return nil, nil
 	}
 
-	value := f(*original)
-	return &value
-}
-
-func toInt64(original int) int64 {
-	return int64(original)
-}
-
-// idForEndpoint returns a unique identifier for an endpoint
-func idForEndpoint(dnsRecord *v1alpha1.DNSRecord, endpoint *v1alpha1.Endpoint) (string, error) {
-	hash := md5.New()
-	if _, err := io.WriteString(hash, fmt.Sprintf("%s/%s@%s", dnsRecord.Name, endpoint.SetIdentifier, endpoint.DNSName)); err != nil {
-		return "", fmt.Errorf("unexpected error creating ID for endpoint %s", endpoint.SetIdentifier)
+	for _, listener := range gw.Spec.Listeners {
+		dnsRecord := &v1alpha1.DNSRecord{}
+		log.V(3).Info("getting dnsrecord", "namespace", dnsPolicy.Namespace)
+		host := listener.Hostname
+		if host == nil {
+			continue
+		}
+		log.V(3).Info("getting dnsrecord", "name", host)
+		if err := r.Client().Get(ctx, client.ObjectKey{Name: string(*host), Namespace: dnsPolicy.Namespace}, dnsRecord); err != nil {
+			if errors.IsNotFound(err) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+		for _, endpoint := range dnsRecord.Spec.Endpoints {
+
+			log.V(1).Info("reconcileHealthChecks: found DNS Record endpoint", "endpoint", endpoint)
+			if endpoint.RecordType == string(v1alpha1.CNAMERecordType) {
+				// if the CNAME is a subdomain of this dnsrecord, skip it
+				if dns.IsSubDomain(dnsRecord.Name, endpoint.Targets[0]) {
+					continue
+				}
+			} else if endpoint.RecordType != string(v1alpha1.ARecordType) {
+				log.V(1).Info("reconcileHealthChecks: not an A record, skipping")
+				continue
+			}
+			for _, target := range endpoint.Targets {
+				log.V(1).Info("reconcileHealthChecks: adding health check for target", "target", target)
+				healthCheck := &v1alpha1.DNSHealthCheckProbe{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fmt.Sprintf("%s-%s", target, listener.Name),
+						Namespace: gw.Namespace,
+						Labels:    dnsRecordLabels(client.ObjectKeyFromObject(gw), client.ObjectKeyFromObject(dnsPolicy)),
+					},
+					Spec: v1alpha1.DNSHealthCheckProbeSpec{
+						Port:              int(listener.Port),
+						Host:              string(*listener.Hostname),
+						Address:           target,
+						Path:              dnsPolicy.Spec.HealthCheck.Endpoint,
+						Protocol:          *dnsPolicy.Spec.HealthCheck.Protocol,
+						Interval:          metav1.Duration{Duration: 60 * time.Second},
+						AdditionalHeaders: dnsPolicy.Spec.HealthCheck.AdditionalHeaders,
+						FailureThreshold:  dnsPolicy.Spec.HealthCheck.FailureThreshold,
+						ExpectedReponses:  dnsPolicy.Spec.HealthCheck.ExpectedResponses,
+					},
+				}
+				healthChecks = append(healthChecks, healthCheck)
+			}
+		}
```

Comment:
@philbrookes I was thinking about this again, and was wondering if it would be better here, rather than processing a gateways listeners (which the DNSrecord reconcile is already doing), to instead just work off the list of DNSRecords that were created by this DNSPolicy?

HealthChecks are more tightly coupled to DNSRecords than the Gateway itself, and since we have just processed the records to create/update/delete them, it seems logical, to me at least, that we could just iterate round them here.

Note: This might be easier once we have these changes in https://github.com/mikenairn/multi-cluster-traffic-controller/commit/655596b5845d8c4ce4ef0b7848846f70b989acf8.

Reply:
for this to work, the dns policy would have to attach to the DNS Record rather than the gateway, otherwise how would the DNS Record reconciler know what configuration to use in the DNS Health Check Probe for each endpoint?

Reply:
You have the policy here, you are doing this as part of the policy reconciliation. I'm just saying you don't need to get records by looking them up based on the listeners of the gateway, the DNSRecord reconcile has already done that, and any difference in the logic in the two could result in different things happening. You could list the dnsrecords in the same way this delete is doing here https://github.com/Kuadrant/multicluster-gateway-controller/pull/379/files#diff-560235894bef952ec7cc558026db8ebfadbab2c1667248e75eb0c575d2ff2e0bR124-R134 and create your probes for each.

Reply:
ah, I see. I thought you were suggesting moving this logic into the DNS Record reconciler. Yes, this makes sense, I'll add it.

Reply:
OK, maybe there are other considerations. I wasn't suggesting we had to change it for this PR to merge though, so we can maybe think on that one.

Reply:
oh no, I take that back, the port is specified in the DNS Policy. I'll add this change now.